test/run_tests.py

```diff
@@ -42,20 +42,87 @@ class Result(Enum):
     SKIPPED = 3
 
 
-def try_decode(binary_object):
+class Completion:
+    def __init__(self):
+        self.lock = threading.Lock()
+        self.condition = threading.Condition(self.lock)
+        self.completed: Dict[int, bool] = dict()
+
+    def report_completion(self, id, success):
+        self.lock.acquire()
+        self.completed[id] = success
+        self.condition.notify_all()
+        self.lock.release()
+
+    def wait_for_completion_get_success(self, id):
+        def is_complete():
+            return id in self.completed
+
+        self.lock.acquire()
+        if not is_complete():
+            self.condition.wait_for(is_complete)
+        success = self.completed[id]
+        self.lock.release()
+        return success
+
+
+class StatusCheck:
+    success: bool
+    message: str
+    stdout: str
+    stderr: str
+
+    def __init__(self, success: bool, message: str, stdout: str, stderr: str):
+        self.success = success
+        self.message = message
+        self.stdout = stdout
+        self.stderr = stderr
+
+
+class DiffCheck:
+    success: bool
+    message: str
+    diff: List[str]
+
+    def __init__(self, success: bool, message: str, diff: List[str]):
+        self.success = success
+        self.message = message
+        self.diff = diff
+
+
+class TestOutcome:
+    id: int
+    result: Result
+    checks: Dict[str, Union[StatusCheck, DiffCheck]]
+    skip_reason: Optional[str]
+
+    def __init__(
+        self,
+        id: int,
+        result: Result,
+        checks: Dict[str, Union[StatusCheck, DiffCheck]],
+        skip_reason: Optional[str] = None,
+    ):
+        self.id = id
+        self.result = result
+        self.checks = checks
+        self.skip_reason = skip_reason
+
+
+def try_decode(binary_object) -> str:
```

Comment:
and the same here - type hint for binary_object?

Reply:
I'm not actually sure what the type of binary object is, I'll look into that one